### PR TITLE
TW-800 Apply new Search inside Chat design

### PR DESCRIPTION
**Pull request by drminh2807**
# Issue
- #800 

# Demo
Screenshot 2023-10-25 at 14 55 59

https://github.com/linagora/twake-on-matrix/assets/12546908/0ef35ad6-ee3a-4c43-bc3a-78d40ea36657

https://github.com/linagora/twake-on-matrix/assets/12546908/69aa7b40-8fa0-4ef2-88f1-ef5c2d0638e2

# Applied design

https://github.com/linagora/twake-on-matrix/assets/12546908/7c076f1b-06c8-4384-9e5c-c3271e075798

# Empty view

https://github.com/linagora/twake-on-matrix/assets/12546908/e031213d-2852-48c4-bdb9-470734a1b845

### CODE REVIEW

File: lib/utils/string_extension.dart [hunk 1]
```diff
@@ -290,4 +290,12 @@ extension StringCasingExtension on String {
     );
     return result;
   }
+
+  String substringToHighlight(String highlightText, {int prefixLength = 0}) {
+    final index = toLowerCase().indexOf(highlightText.toLowerCase());
```
Comment (sherlockvn) [inline lib/utils/string_extension.dart:295]: should check the prefixLength < 0

Comment (sherlockvn) [reply lib/utils/string_extension.dart:295]: index also possible = -1, should check that case also

Comment (sherlockvn) [reply lib/utils/string_extension.dart:295]: i don't know what about in the case have icons ?

**Comment by hoangdat** [comment]
style is not the same with design

image

File: lib/pages/chat/events/message_download_content.dart [hunk 2]
```diff
@@ -37,9 +33,11 @@ class MessageDownloadContent extends StatelessWidget {
       'filename: $filename, filetype: $filetype, sizeString: $sizeString, content: ${event.content}',
     );
     return InkWell(
-      onTap: () async {
-        onFileTapped(event);
-      },
+      onTap: onFileTapped != null
+          ? () async {
```
Comment (imGok) [inline lib/pages/chat/events/message_download_content.dart:37]: why is async needed

File: lib/pages/chat_adaptive_scaffold/chat_adaptive_scaffold.dart [hunk 3]
```diff
@@ -21,11 +24,20 @@ class ChatAdaptiveScaffold extends StatefulWidget {
...
-class _ChatAdaptiveScaffoldState extends State<ChatAdaptiveScaffold> {
+class ChatAdaptiveScaffoldController extends State<ChatAdaptiveScaffold> {
   final showRightPanelNotifier = ValueNotifier(false);
+  final jumpToEventIdStream = StreamController<EventId>.broadcast();
+
+  void jumpToEventId(String eventId) {
+    jumpToEventIdStream.sink.add(eventId);
+  }
+
+  Future togglePage() async {
```
Comment (imGok) [inline lib/pages/chat_adaptive_scaffold/chat_adaptive_scaffold.dart:38]: same, why you need a future here ?

File: lib/pages/chat_adaptive_scaffold/chat_adaptive_scaffold.dart [hunk 4]
```diff
@@ -43,13 +55,19 @@ class _ChatAdaptiveScaffoldState extends State<ChatAdaptiveScaffold> {
...
-                        ChatSearch(
-                          roomId: widget.roomId,
-                          onBack: showRightPanelNotifier.toggle,
+                        MaterialPage(
+                          child: ChatSearch(
+                            roomId: widget.roomId,
+                            onBack: togglePage,
+                            jumpToEventId: jumpToEventId,
+                            isInStack: true,
+                          ),
```
Comment (nqhhdev) [inline lib/pages/chat_adaptive_scaffold/chat_adaptive_scaffold.dart:70]: Can you explain it?

File: lib/pages/chat_search/chat_search.dart [hunk 5]
```diff
@@ -26,8 +41,57 @@ class ChatSearchController extends State<ChatSearch> {
 
   final textEditingController = TextEditingController();
 
+  final inputFocus = FocusNode();
+
   final debouncer = Debouncer(_debouncerDuration, initialValue: '');
 
+  SameTypeEventsBuilderController? eventsController;
```
[the remark this answers is not in the document]
Comment (drminh2807) [reply lib/pages/chat_search/chat_search.dart:48]: late can make crash when call before initialize
so to void crash, use nullable

File: lib/pages/chat_adaptive_scaffold/chat_adaptive_scaffold.dart (hunk 4 above)
Comment (drminh2807) [reply lib/pages/chat_adaptive_scaffold/chat_adaptive_scaffold.dart:70]: push new screen without `push`, use `pages` array instead

File: lib/utils/string_extension.dart (hunk 1 above)
Comment (drminh2807) [reply lib/utils/string_extension.dart:296]: done

File: lib/pages/chat/events/message_download_content.dart (hunk 2 above)
Comment (drminh2807) [reply lib/pages/chat/events/message_download_content.dart]: fixed

File: lib/pages/chat_adaptive_scaffold/chat_adaptive_scaffold.dart (hunk 3 above)
Comment (drminh2807) [reply lib/pages/chat_adaptive_scaffold/chat_adaptive_scaffold.dart]: fixed

File: lib/pages/chat/chat.dart [hunk 6]
```diff
@@ -303,13 +304,9 @@ class ChatController extends State<Chat>
   void initState() {
     keyboardVisibilityController.onChange.listen(_keyboardListener);
     scrollController.addListener(_updateScrollController);
+    widget.jumpToEventIdStream?.stream.listen(scrollToEventId);
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:307]: Please don't forget to cancel this `stream subscription`

**Review by hoangdat — COMMENTED** [review]
fix toggle problem:
- open search by icon in app bar
- click to open chat info
- click on search icon in chat info
- issue: search is close -> expect: search still open

**Comment by drminh2807** [comment]
> fix toggle problem:
> 
> * open search by icon in app bar
> * click to open chat info
> * click on search icon in chat info
> * issue: search is close -> expect: search still open

done

File: lib/pages/chat/chat.dart (hunk 6 above)
Comment (drminh2807) [reply lib/pages/chat/chat.dart:307]: done

**Review by hoangdat — APPROVED** [review]

File: lib/pages/chat_adaptive_scaffold/chat_adaptive_scaffold.dart [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    this.roomName,
+  }) : super(key: key);
+
+  @override
+  State<ChatAdaptiveScaffold> createState() => ChatAdaptiveScaffoldController();
+}
+
+class ChatAdaptiveScaffoldController extends State<ChatAdaptiveScaffold> {
+  final showRightPanelNotifier = ValueNotifier(false);
+  final jumpToEventIdStream = StreamController<EventId>.broadcast();
```
Comment (nqhhdev) [inline lib/pages/chat_adaptive_scaffold/chat_adaptive_scaffold.dart:32]: Don't forgot dispose `ValueNotifier` and cancel  `StreamController`

File: lib/pages/chat_adaptive_scaffold/chat_adaptive_scaffold.dart [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+                  ),
+                ),
+                const WidthPlatformBreakpoint(
+                  begin: breakpoint,
+                ): SlotLayout.from(
+                  key: AppAdaptiveScaffold.breakpointWebAndDesktopKey,
+                  builder: (_) => Padding(
+                    padding: showRightPanel
+                        ? const EdgeInsetsDirectional.only(end: 16)
+                        : EdgeInsets.zero,
```
Comment (nqhhdev) [inline lib/pages/chat_adaptive_scaffold/chat_adaptive_scaffold.dart:86]: Move to style

File: lib/pages/chat_adaptive_scaffold/chat_adaptive_scaffold.dart (hunk 7 above)
Comment (drminh2807) [reply lib/pages/chat_adaptive_scaffold/chat_adaptive_scaffold.dart:33]: done

File: lib/pages/chat_adaptive_scaffold/chat_adaptive_scaffold.dart (hunk 8 above)
Comment (drminh2807) [reply lib/pages/chat_adaptive_scaffold/chat_adaptive_scaffold.dart]: done

**Review by nqhhdev — APPROVED** [review]

**Comment by sherlockvn** [comment]
- i think maybe change the empty view is like search in contacts, instead of 404. wdyt? @hoangdat 
- Can you check again with responsive web @drminh2807

**Review by sherlockvn — APPROVED** [review]

**Pull request merged**